File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val = 0, _next = nil)
#     @val = val
#     @next = _next
#   end
# end
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {ListNode} head
# @return {TreeNode}
def helper(lb, ub, lst)
  return nil if lb == ub
  mid = (lb+ub) / 2
  return TreeNode.new(lst[mid].val, helper(lb, mid, lst), helper(mid+1, ub, lst))
end

def sorted_list_to_bst(head)
  lst, curr = [], head
  while curr
    lst << curr
    curr = curr.next
  end
  
  return helper(0, lst.length, lst)
end

File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *   int val;
 *   ListNode next;
 *   ListNode() {}
 *   ListNode(int val) { this.val = val; }
 *   ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *   int val;
 *   TreeNode left;
 *   TreeNode right;
 *   TreeNode() {}
 *   TreeNode(int val) { this.val = val; }
 *   TreeNode(int val, TreeNode left, TreeNode right) {
 *     this.val = val;
 *     this.left = left;
 *     this.right = right;
 *   }
 * }
 */
class Solution {
  public TreeNode helper(int lb, int ub, List<ListNode> lst) {
    if (lb == ub) return null;
    int mid = (lb + ub) / 2;
    return new TreeNode(
      lst.get(mid).val,
      helper(lb, mid, lst),
      helper(mid + 1, ub, lst)
    );
  }
  
  public TreeNode sortedListToBST(ListNode head) {
    List<ListNode> lst = new ArrayList<ListNode>();
    ListNode curr = head;
    while (curr != null) {
      lst.add(curr);
      curr = curr.next;
    }
    
    return helper(0, lst.size(), lst);
  }
}